python/cudf/cudf/tests/test_concat.py

```diff
@@ -1203,3 +1203,14 @@ def test_concat_join_empty_dataframes_axis_1(
     assert_eq(
         expected, actual, check_index_type=False, check_column_type=False
     )
+
+
+def test_concat_preserve_order():
```

Comment:
Do we need more tests/revisal of previous tests to cover the bugs fixed in this PR?

Reply:
Bugs 1 and 2 are both already detected by other test code paths now that 3 and 5 are fixed (this new test is for 5). Bug 4 is more an implementation detail than an API change, so I didn't bother to add one. I can not be lazy and create a test for that case (using an `Index` object to specify the columns) if you think it's worth it.

Reply:
We shouldn't add tests for implementation detail though.

Reply:
Right, I agree. It's just that this implementation detail has an API implication. Users can call `pd.concat(..., columns=pd.Index(list_of_column_names))`, which should be equivalent to `pd.concat(..., columns=list_of_column_names)`. However, without the fix I did for bug 4, the first case will error. Checking `if []` is valid Python, it will evaluate to `False` if the list is empty and `True` otherwise, but `pd.Index` doesn't support the same check because it behaves more like `numpy.ndarray` where you have to do elementwise true/false checks or something like `all(index == 0)`. Therefore, we can't just do `if columns`, we have to do `if len(columns) != 0`. The relevant test to add would be to do `pd.concat` with a `pd.Index` or `cudf.Index` as the `columns` object. I'll just not be lazy and add the test :)